### ISSUE TITLE
The diagnostics of `sumneko_lua` does not appear after `:e`

### ISSUE THREAD

**Issue by nyngwang** [neovim/nvim-lspconfig#1701]
### Description

As title. How to make things clear? This is the DEMO:

(This might be helpful but I'm not sure: https://github.com/neovim/nvim-lspconfig/issues/1192)

https://user-images.githubusercontent.com/24765272/153084238-dc454d4b-599c-4b98-be18-93bc8dd54fa0.mov

### Neovim version

NVIM v0.7.0-dev+1036-gbcea73299
Build type: Release
LuaJIT 2.1.0-beta3

### Nvim-lspconfig version

2008c5cebf2b84c5e5f8a566480b022ab2e7ebab

### Operating system and version

macOS Big Sur 11.6.2

### Affected language servers

sumneko_lua

### Steps to reproduce

As shown in the DEMO.

### Actual behavior

As shown in the DEMO.

### Expected behavior

As shown in the DEMO.

### Minimal config

```Lua
local runtime_path = '~/.local/share/nvim/lsp_servers/sumneko_lua/extension/server/bin/lua-language-server'

M = {
  capabilities = require('cmp_nvim_lsp').update_capabilities(vim.lsp.protocol.make_client_capabilities()),
  autostart = true,
  on_attach = function (...)
    require('lsp_config.common_setup')(...)
  end,
  settings = {
    Lua = {
      runtime = {
        version = 'LuaJIT',
        path = runtime_path
      },
      diagnostics = {
        globals = { 'vim' }
      }
    }
  }
}

return M
```

### LSP log

```

[WARN][2022-02-09 06:09:02] .../lua/vim/lsp.lua:812	"server_request: no handler found for"	"workspace/semanticTokens/refresh"
[WARN][2022-02-09 06:09:02] .../lua/vim/lsp.lua:812	"server_request: no handler found for"	"workspace/semanticTokens/refresh"
[WARN][2022-02-09 06:09:02] .../lua/vim/lsp.lua:812	"server_request: no handler found for"	"workspace/semanticTokens/refresh"
[WARN][2022-02-09 06:09:02] .../lua/vim/lsp.lua:812	"server_request: no handler found for"	"workspace/semanticTokens/refresh"
[START][2022-02-09 06:09:29] LSP logging initiated
[WARN][2022-02-09 06:09:30] .../lua/vim/lsp.lua:812	"server_request: no handler found for"	"workspace/semanticTokens/refresh"
[WARN][2022-02-09 06:09:30] .../lua/vim/lsp.lua:812	"server_request: no handler found for"	"workspace/semanticTokens/refresh"
[START][2022-02-09 06:14:08] LSP logging initiated
[WARN][2022-02-09 06:14:13] .../lua/vim/lsp.lua:812	"server_request: no handler found for"	"workspace/semanticTokens/refresh"
[WARN][2022-02-09 06:14:13] .../lua/vim/lsp.lua:812	"server_request: no handler found for"	"workspace/semanticTokens/refresh"
```

**Comment by mjlbach** [comment]
You can file an issue upstream. What ":e" does is effectively close/reopen the buffer. Servers by convention send diagnostics when text document's are opened, in this case `lua-language-server` only sends diagnostics when the buffer is reopened (`textDocument/didClose` followed by `textDocument/didOpen`)  *and* the first `textDocument/didChange` notification is sent.

```lua
[DEBUG][2022-02-08 14:21:52] .../vim/lsp/rpc.lua:347	"rpc.send"	{
  jsonrpc = "2.0",
  method = "textDocument/didClose",
  params = {
    textDocument = {
      uri = "file:///Users/michael/Repositories/nvim-lspconfig/test/minimal_init.lua"
    }
  }
}
[DEBUG][2022-02-08 14:21:52] .../vim/lsp/rpc.lua:347	"rpc.send"	{
  jsonrpc = "2.0",
  method = "textDocument/didOpen",
  params = {
    textDocument = {
      languageId = "lua",
      text = "local on_windows = vim.loop.os_uname().version:match 'Windows'\n\nlocal function join_paths(...)\n  local path_sep = on_windows and '\\\\' or '/'\n  local result = table.concat({ ... }, path_sep)\n  return result\nend\n\nvim.cmd [[set runtimepath=$VIMRUNTIME]]\n\nlocal temp_dir = vim.loop.os_getenv 'TEMP' or '/tmp'\n\nvim.cmd('set packpath=' .. join_paths(temp_dir, 'nvim', 'site'))\n\nlocal package_root = join_paths(temp_dir, 'nvim', 'site', 'pack')\nlocal install_path = join_paths(package_root, 'packer', 'start', 'packer.nvim')\nlocal compile_path = join_paths(install_path, 'plugin', 'packer_compiled.lua')\n\nlocal function load_plugins()\n  require('packer').startup {\n    {\n      'wbthomason/packer.nvim',\n      'neovim/nvim-lspconfig',\n    },\n    config = {\n      package_root = package_root,\n      compile_path = compile_path,\n    },\n  }\nend\n\n_G.load_config = function()\n  vim.lsp.set_log_level 'trace'\n  if vim.fn.has 'nvim-0.5.1' == 1 then\n    require('vim.lsp.log').set_format_func(vim.inspect)\n  end\n  local nvim_lsp = require 'lspconfig'\n  local on_attach = function(_, bufnr)\n    local function buf_set_keymap(...)\n      vim.api.nvim_buf_set_keymap(bufnr, ...)\n    end\n    local function buf_set_option(...)\n      vim.api.nvim_buf_set_option(bufnr, ...)\n    end\n\n    buf_set_option('omnifunc', 'v:lua.vim.lsp.omnifunc')\n\n    -- Mappings.\n    local opts = { noremap = true, silent = true }\n    buf_set_keymap('n', 'gD', '<Cmd>lua vim.lsp.buf.declaration()<CR>', opts)\n    buf_set_keymap('n', 'gd', '<Cmd>lua vim.lsp.buf.definition()<CR>', opts)\n    buf_set_keymap('n', 'K', '<Cmd>lua vim.lsp.buf.hover()<CR>', opts)\n    buf_set_keymap('n', 'gi', '<cmd>lua vim.lsp.buf.implementation()<CR>', opts)\n    buf_set_keymap('n', '<C-k>', '<cmd>lua vim.lsp.buf.signature_help()<CR>', opts)\n    buf_set_keymap('n', '<space>wa', '<cmd>lua vim.lsp.buf.add_workspace_folder()<CR>', opts)\n    buf_set_keymap('n', '<space>wr', '<cmd>lua vim.lsp.buf.remove_workspace_folder()<CR>', opts)\n    buf_set_keymap('n', '<space>wl', '<cmd>lua print(vim.inspect(vim.lsp.buf.list_workspace_folders()))<CR>', opts)\n    buf_set_keymap('n', '<space>D', '<cmd>lua vim.lsp.buf.type_definition()<CR>', opts)\n    buf_set_keymap('n', '<space>rn', '<cmd>lua vim.lsp.buf.rename()<CR>', opts)\n    buf_set_keymap('n', 'gr', '<cmd>lua vim.lsp.buf.references()<CR>', opts)\n    buf_set_keymap('n', '<space>e', '<cmd>lua vim.lsp.diagnostic.show_line_diagnostics()<CR>', opts)\n    buf_set_keymap('n', '[d', '<cmd>lua vim.lsp.diagnostic.goto_prev()<CR>', opts)\n    buf_set_keymap('n', ']d', '<cmd>lua vim.lsp.diagnostic.goto_next()<CR>', opts)\n    buf_set_keymap('n', '<space>q', '<cmd>lua vim.lsp.diagnostic.set_loclist()<CR>', opts)\n  end\n\n  -- Add the server that troubles you here\n  local name = 'sumneko_lua'\n  if not name then\n    print 'You have not defined a server name, please edit minimal_init.lua'\n  end\n  if not nvim_lsp[name].document_config.default_config.cmd and not cmd then\n    print [[You have not defined a server default cmd for a server\n      that requires it please edit minimal_init.lua]]\n  end\n\n  nvim_lsp[name].setup {\n    cmd = cmd,\n    on_attach = on_attach,\n  }\n\n  print [[You can find your log at $HOME/.cache/nvim/lsp.log. Please paste in a github issue under a details tag as described in the issue template.]]\nend\n\nif vim.fn.isdirectory(install_path) == 0 then\n  vim.fn.system { 'git', 'clone', 'https://github.com/wbthomason/packer.nvim', install_path }\n  load_plugins()\n  require('packer').sync()\n  vim.cmd [[autocmd User PackerComplete ++once lua load_config()]]\nelse\n  load_plugins()\n  require('packer').sync()\n  _G.load_config()\nend\n",
      uri = "file:///Users/michael/Repositories/nvim-lspconfig/test/minimal_init.lua",
      version = 0
    }
  }
}
[DEBUG][2022-02-08 14:21:53] .../vim/lsp/rpc.lua:347	"rpc.send"	{
  jsonrpc = "2.0",
  method = "textDocument/didChange",
  params = {
    contentChanges = { {
        range = {
          end = {
            character = 0,
            line = 13
          },
          start = {
            character = 0,
            line = 13
          }
        },
        rangeLength = 0,
        text = "e"
      } },
    textDocument = {
      uri = "file:///Users/michael/Repositories/nvim-lspconfig/test/minimal_init.lua",
      version = 6
    }
  }
}
[DEBUG][2022-02-08 14:21:53] .../vim/lsp/rpc.lua:454	"rpc.receive"	{
  jsonrpc = "2.0",
  method = "$/status/report",
  params = {
    text = "😺Lua",
    tooltip = "Workspace   : /Users/michael/Repositories/nvim-lspconfig\nCached files: 147/146\nMemory usage: 15M"
  }
}
[TRACE][2022-02-08 14:21:53] .../lua/vim/lsp.lua:792	"notification"	"$/status/report"	{
  text = "😺Lua",
  tooltip = "Workspace   : /Users/michael/Repositories/nvim-lspconfig\nCached files: 147/146\nMemory usage: 15M"
}
[DEBUG][2022-02-08 14:21:53] .../vim/lsp/rpc.lua:454	"rpc.receive"	{
  jsonrpc = "2.0",
  method = "textDocument/publishDiagnostics",
  params = {
    diagnostics = { {
        code = "exp-in-action",
        data = "syntax",
        message = "Unexpected <exp> .",
        range = {
          end = {
            character = 1,
            line = 13
          },
          start = {
            character = 0,
            line = 13
          }
        },
        severity = 1,
        source = "Lua Syntax Check."
      }, {
        code = "undefined-global",
        message = "Undefined global `vim`.",
        range = {
          end = {
            character = 3,
            line = 8
          },
          start = {
            character = 0,
            line = 8
          }
        },
        severity = 2,
        source = "Lua Diagnostics."
      }, {
        code = "undefined-global",
        message = "Undefined global `vim`.",
        range = {
          end = {
            character = 3,
            line = 12
          },
          start = {
            character = 0,
            line = 12
          }
        },
        severity = 2,
        source = "Lua Diagnostics."
      }, {
        code = "undefined-global",
        message = "Undefined global `vim`.",
        range = {
          end = {
            character = 5,
            line = 88
          },
          start = {
            character = 2,
            line = 88
          }
        },
        severity = 2,
        source = "Lua Diagnostics."
      }, {
        code = "undefined-global",
        message = "Undefined global `vim`.",
        range = {
          end = {
            character = 20,
            line = 10
          },
          start = {
            character = 17,
            line = 10
          }
        },
        severity = 2,
        source = "Lua Diagnostics."
      }, {
        code = "undefined-global",
        message = "Undefined global `vim`.",
        range = {
          end = {
            character = 5,
            line = 32
          },
          start = {
            character = 2,
            line = 32
          }
        },
        severity = 2,
        source = "Lua Diagnostics."
      }, {
        code = "undefined-global",
        message = "Undefined global `vim`.",
        range = {
          end = {
            character = 5,
            line = 85
          },
          start = {
            character = 2,
            line = 85
          }
        },
        severity = 2,
        source = "Lua Diagnostics."
      }, {
        code = "undefined-global",
        message = "Undefined global `cmd`.",
        range = {
          end = {
            character = 70,
            line = 71
          },
          start = {
            character = 67,
            line = 71
          }
        },
        severity = 2,
        source = "Lua Diagnostics."
      }, {
        code = "undefined-global",
        message = "Undefined global `cmd`.",
        range = {
          end = {
            character = 13,
            line = 77
          },
          start = {
            character = 10,
            line = 77
          }
        },
        severity = 2,
        source = "Lua Diagnostics."
      }, {
        code = "undefined-global",
        message = "Undefined global `vim`.",
        range = {
          end = {
            character = 6,
            line = 84
          },
          start = {
            character = 3,
            line = 84
          }
        },
        severity = 2,
        source = "Lua Diagnostics."
      }, {
        code = "undefined-global",
        message = "Undefined global `vim`.",
        range = {
          end = {
            character = 46,
            line = 34
          },
          start = {
            character = 43,
            line = 34
          }
        },
        severity = 2,
        source = "Lua Diagnostics."
      }, {
        code = "undefined-global",
        message = "Undefined global `vim`.",
        range = {
          end = {
            character = 22,
            line = 0
          },
          start = {
            character = 19,
            line = 0
          }
        },
        severity = 2,
        source = "Lua Diagnostics."
      }, {
        code = "undefined-global",
        message = "Undefined global `vim`.",
        range = {
          end = {
            character = 8,
            line = 33
          },
          start = {
            character = 5,
            line = 33
          }
        },
        severity = 2,
        source = "Lua Diagnostics."
      }, {
        code = "undefined-global",
        message = "Undefined global `vim`.",
        range = {
          end = {
            character = 9,
            line = 39
          },
          start = {
            character = 6,
            line = 39
          }
        },
        severity = 2,
        source = "Lua Diagnostics."
      }, {
        code = "undefined-global",
        message = "Undefined global `vim`.",
        range = {
          end = {
            character = 9,
            line = 42
          },
          start = {
            character = 6,
            line = 42
          }
        },
        severity = 2,
        source = "Lua Diagnostics."
      } },
    uri = "file:///Users/michael/Repositories/nvim-lspconfig/test/minimal_init.lua",
    version = 6
  }
}
[TRACE][2022-02-08 14:21:53] .../lua/vim/lsp.lua:792	"notification"	"textDocument/publishDiagnostics"	{
  diagnostics = { {
      code = "exp-in-action",
      data = "syntax",
      message = "Unexpected <exp> .",
      range = {
        end = {
          character = 1,
          line = 13
        },
        start = {
          character = 0,
          line = 13
        }
      },
      severity = 1,
      source = "Lua Syntax Check."
    }, {
      code = "undefined-global",
      message = "Undefined global `vim`.",
      range = {
        end = {
          character = 3,
          line = 8
        },
        start = {
          character = 0,
          line = 8
        }
      },
      severity = 2,
      source = "Lua Diagnostics."
    }, {
      code = "undefined-global",
      message = "Undefined global `vim`.",
      range = {
        end = {
          character = 3,
          line = 12
        },
        start = {
          character = 0,
          line = 12
        }
      },
      severity = 2,
      source = "Lua Diagnostics."
    }, {
      code = "undefined-global",
      message = "Undefined global `vim`.",
      range = {
        end = {
          character = 5,
          line = 88
        },
        start = {
          character = 2,
          line = 88
        }
      },
      severity = 2,
      source = "Lua Diagnostics."
    }, {
      code = "undefined-global",
      message = "Undefined global `vim`.",
      range = {
        end = {
          character = 20,
          line = 10
        },
        start = {
          character = 17,
          line = 10
        }
      },
      severity = 2,
      source = "Lua Diagnostics."
    }, {
      code = "undefined-global",
      message = "Undefined global `vim`.",
      range = {
        end = {
          character = 5,
          line = 32
        },
        start = {
          character = 2,
          line = 32
        }
      },
      severity = 2,
      source = "Lua Diagnostics."
    }, {
      code = "undefined-global",
      message = "Undefined global `vim`.",
      range = {
        end = {
          character = 5,
          line = 85
        },
        start = {
          character = 2,
          line = 85
        }
      },
      severity = 2,
      source = "Lua Diagnostics."
    }, {
      code = "undefined-global",
      message = "Undefined global `cmd`.",
      range = {
        end = {
          character = 70,
          line = 71
        },
        start = {
          character = 67,
          line = 71
        }
      },
      severity = 2,
      source = "Lua Diagnostics."
    }, {
      code = "undefined-global",
      message = "Undefined global `cmd`.",
      range = {
        end = {
          character = 13,
          line = 77
        },
        start = {
          character = 10,
          line = 77
        }
      },
      severity = 2,
      source = "Lua Diagnostics."
    }, {
      code = "undefined-global",
      message = "Undefined global `vim`.",
      range = {
        end = {
          character = 6,
          line = 84
        },
        start = {
          character = 3,
          line = 84
        }
      },
      severity = 2,
      source = "Lua Diagnostics."
    }, {
      code = "undefined-global",
      message = "Undefined global `vim`.",
      range = {
        end = {
          character = 46,
          line = 34
        },
        start = {
          character = 43,
          line = 34
        }
      },
      severity = 2,
      source = "Lua Diagnostics."
    }, {
      code = "undefined-global",
      message = "Undefined global `vim`.",
      range = {
        end = {
          character = 22,
          line = 0
        },
        start = {
          character = 19,
          line = 0
        }
      },
      severity = 2,
      source = "Lua Diagnostics."
    }, {
      code = "undefined-global",
      message = "Undefined global `vim`.",
      range = {
        end = {
          character = 8,
          line = 33
        },
        start = {
          character = 5,
          line = 33
        }
      },
      severity = 2,
      source = "Lua Diagnostics."
    }, {
      code = "undefined-global",
      message = "Undefined global `vim`.",
      range = {
        end = {
          character = 9,
          line = 39
        },
        start = {
          character = 6,
          line = 39
        }
      },
      severity = 2,
      source = "Lua Diagnostics."
    }, {
      code = "undefined-global",
      message = "Undefined global `vim`.",
      range = {
        end = {
          character = 9,
          line = 42
        },
        start = {
          character = 6,
          line = 42
        }
      },
      severity = 2,
      source = "Lua Diagnostics."
    } },
  uri = "file:///Users/michael/Repositories/nvim-lspconfig/test/minimal_init.lua",
  version = 6
}
```

**Comment by nyngwang** [comment]
@mjlbach: Wait, but in the last part of the DEMO, the diagnostics are still shown?

**Comment by mjlbach** [comment]
Yes, did you see the RPC log? When you send a text edit after you hit `:e` (as you do in the video) the diagnostics are sent via notification from the server. I reproduced the exact same thing locally. A similar issue was present in gopls that I reported: https://github.com/golang/go/issues/50267

In the future, I find videos hard to follow. And you should also reproduce/upload a minimal config, not a section of your configuration.

**Comment by nyngwang** [comment]
OK, sorry... I'm very bad about how to create a minimal config, so I thought DEMO is the most concise way for me to express what happened.

**Comment by nyngwang** [comment]
@mjlbach: Thanks for your help! Since I don't understand this topic well, I will not reply to that thread further. But if some debugging logs are needed feel free to tag me!

**Comment by adrian5** [comment]
@nyngwang Sorry to barge into this closed topic.

I noticed that you hardcode `runtime_path`. Is that simply the path to the server script/exectuable? The examples shown by lspconfig create a messy array of paths instead, which I assumed was somehow necessary. Can I just omit that if I know where the server executable is located?

```lua
-- Yikes
local runtime_path = vim.split(package.path, ';')
table.insert(runtime_path, "lua/?.lua")
table.insert(runtime_path, "lua/?/init.lua")
```

**Comment by nyngwang** [comment]
To @adrian5: Hi,

> [...] Is that simply the path to the server script/exectuable? [...]

Yes, macOS seems to always use that path to install my neovim stuff. It worked, so I left it there.

> [...] Can I just omit that if I know where the server executable is located?

Just try it? I haven't encountered any trouble with that.

Sincerely,
Rain
